stat/test/wilcoxon_hyp.go

```diff
@@ -0,0 +1,182 @@
+	Position int
+	Value    float64
+}
+
+func ensureDataConformance(x, y []float64) {
+	if len(x) == 0 || len(y) == 0 {
+		panic(badLengthIsZero)
+	}
+	if len(x) != len(y) {
+		panic(badLengthMismatch)
+	}
+}
+
+func calculateAbsDifference(dst, x, y []float64) {
+	for index, xVal := range x {
+		dst[index] = math.Abs(y[index] - xVal)
+	}
+}
+
+func rank(out, in []float64) int {
+	tieAdjust := 0
+	lengthOfIn := len(in)
+	pairStructArr := make([]pairStruct, lengthOfIn)
+	for index, inVal := range in {
+		pairStructArr[index] = pairStruct{
+			Position: index,
+			Value:    inVal,
+		}
+	}
+	sort.Slice(pairStructArr, func(i, j int) bool {
+		return pairStructArr[i].Value < pairStructArr[j].Value
+	})
+
+	mpOut := make(map[float64]int, len(out))   // tracks the rank
+	mpCount := make(map[float64]int, len(out)) // tracks the count
+	tmpRank := make(map[float64]struct{})      // tracks the tied ranks
+	rank := 1
+	for i := 0; i < lengthOfIn; i++ {
+		val := pairStructArr[i].Value
+		if val != 0 {
+			if _, ok := mpOut[val]; !ok {
+				mpOut[val] = rank
+			} else {
+				mpOut[val] += rank
+			}
+			if _, ok := mpCount[val]; !ok {
+				mpCount[val] = 1
+			} else {
+				mpCount[val] += 1
+			}
+			rank++
+		}
+	}
+
+	var pos int
+	var val float64
+	for i := 0; i < lengthOfIn; i++ {
+		pos = pairStructArr[i].Position
+		val = pairStructArr[i].Value
+		if val != 0 {
+			out[pos] = float64(mpOut[val]) / float64(mpCount[val])
+		}
+		if mpCount[val] > 1 {
+			if _, ok := tmpRank[val]; !ok {
+				tmpRank[val] = struct{}{}
+			}
+		}
+	}
+	for val := range tmpRank {
+		tieAdjust += (mpCount[val] * mpCount[val] * mpCount[val]) - (mpCount[val])
+
+	}
+	return tieAdjust
+}
+
+func wilCoxonSignedRankTest(x, y []float64) (float64, int, int) {
+	z := make([]float64, len(x))
+	absZ := make([]float64, len(x))
+	floats.SubTo(z, x, y)
+
+	allZero := true
+	for _, val := range z {
+		if val != 0 {
+			allZero = false
+			break
+		}
+	}
+	if allZero {
+		panic(badAllElementsAreSame)
+	}
+
+	calculateAbsDifference(absZ, x, y)
+	ranks := make([]float64, len(x))
+	tieAdj := rank(ranks, absZ)
+	n := len(x)
+
+	WPlus := 0.0
+	WMinus := 0.0
+	for index, rank := range ranks {
+		if z[index] > 0 {
+			WPlus += rank
+		} else if z[index] == 0 {
+			n--
+		}
+	}
+	WMinus = (float64(n*(n+1)) / 2.0) - WPlus
+	return math.Max(WMinus, WPlus), n, tieAdj
+}
+
+func calculateExactPValue(Wmax float64, NZ int) float64 {
+	m := 1 << NZ
+	largerRankSums := 0
+
+	for i := 0; i < m; i++ {
+		rankSum := 0
+		// Generate all possible rank sums.
+		for j := 0; j < NZ; j++ {
+			// (i >> j) & 1 extract i's j-th bit from the right
+			if ((i >> j) & 1) == 1 {
+				rankSum += j + 1
+			}
+		}
+		if float64(rankSum) >= Wmax {
+			largerRankSums++
+		}
+	}
+
+	// largerRankSums / m gives the one-sided p-value, so it's multiplied
+	// with 2 to get the two-sided p-value.
+	return 2 * (float64(largerRankSums) / float64(m))
+}
+
+func calculateAsymptoticPValue(Wmin float64, NZ, tieAdj int) float64 {
```

Comment:
I choose 30 because it seemed to give good balance between runtime of the function (because it is a heavy computation to calculate exact p value everytime). This is my reference, https://web.archive.org/web/20070322182510/http://www.fon.hum.uva.nl/Service/Statistics/Signed_Rank_Test.html

Reply:
Thanks. Would you mind adding this link to the documentation?